misc/license_headers.py

```diff
@@ -56,88 +63,128 @@ def check_header(cls, file: Path) -> bool:
     def add_header(cls, file: Path) -> bool:
         with open(file, "r", encoding="utf-8") as fd:
             shabang_line, header_line = extract_shabang_and_header_lines(fd)
-            match = cls.HEADER_RE.match(header_line.strip())
-            if not match:
-                print(f"Add missing {cls.NAME} header", file)
-                updated_data = f"{shabang_line}{cls.HEADER}\n\n{header_line}{fd.read()}"
-                file.write_text(updated_data)
-                return True
-
-            elif match["year"] != THIS_YEAR:
-                print("Correct copyright year", file)
-                updated_data = f"{shabang_line}{cls.HEADER}\n{fd.read()}"
-                file.write_text(updated_data)
+            expected_license_line = cls.generate_license_line()
+            if header_line != expected_license_line:
+                print(f"Add missing {cls.SPDX_ID} header: {file}")
+                updated_data = f"{shabang_line}{expected_license_line}\n{header_line}{fd.read()}"
+                file.write_text(updated_data, encoding="utf-8")
                 return True
 
         return False
 
+    @classmethod
+    def remove_header(cls, file: Path) -> bool:
+        with open(file, "r", encoding="utf-8") as fd:
+            lines = []
+            need_rewrite = False
+            for line in fd.readlines():
```

Comment:
Should only check for a header for the first few lines, not the whole file. It would cause a problem if, for example, you were to launch the script on this very file since the function `is_possible_license_line` would itself be detected as a license header line and be removed.

Reply:
> you were to launch the script on this very file [...]

There is a special case in the file filter to handle this ;-)

```python
    # Special case for ourself given we contain the license headers in the source code !
    re.compile(r"^misc/license_headers.py$"): SkipLicenser,
```

For the rest of the codebase, `is_possible_license_line` checks for `Parsec Cloud (https://parsec.cloud) Copyright (c)` which is highly unlikely to be present (I've tried it on the codebase and only headers are removed)

The thing is this `remove` command is a helper to modify the header line (like what we are going to do just before merging this PR ^^), it is not expected to be run as part of the CI or anything that will go blindly on the codebase